src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs

```diff
@@ -135,6 +135,27 @@ protected override void CheckInterfaces(DiagnosticBag diagnostics)
                 {
                     @interface.CheckAllConstraints(conversions, location, diagnostics);
                 }
+
+                // we check direct base interfaces that have the InternalImplementationOnly attribute,
+                // giving an error when such an attributed interface appears in the interface list.
+                foreach (var baseInterface in this.InterfacesNoUseSiteDiagnostics(null))
+                {
+                    EnsureExtensible(baseInterface, location, diagnostics);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Produce an error if the given base interface may not be extended by this type due to the
+        /// presence of the attribute System.Runtime.CompilerServices.InternalImplementationOnlyAttribute
+        /// </summary>
+        private void EnsureExtensible(NamedTypeSymbol baseInterface, SourceLocation location, DiagnosticBag diagnostics)
+        {
+            if (this.ContainingAssembly != baseInterface.ContainingAssembly &&
+                baseInterface.HasInternalImplementationOnlyAttribute &&
+                !this.ContainingAssembly.HasInternalAccessTo(baseInterface.ContainingAssembly))
```

Comment:
I placed the tests in the order that I expected would minimize the total execution time of this code in a typical compilation.